iceoryx_posh/test/moduletests/test_popo_subscriber_port.cpp

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/iceoryx_posh_types.hpp"
+#include "iceoryx_posh/internal/mepoo/memory_manager.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.hpp"
+#include "iceoryx_posh/internal/popo/ports/subscriber_port_data.hpp"
+#include "iceoryx_posh/internal/popo/ports/subscriber_port_multi_producer.hpp"
+#include "iceoryx_posh/internal/popo/ports/subscriber_port_single_producer.hpp"
+#include "iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp"
+
+#include "iceoryx_posh/mepoo/mepoo_config.hpp"
+#include "iceoryx_utils/error_handling/error_handling.hpp"
+#include "iceoryx_utils/internal/posix_wrapper/shared_memory_object/allocator.hpp"
+#include "test.hpp"
+
+#include <memory>
+
+using namespace ::testing;
+
+class SubscriberPort_test : public Test
+{
+  protected:
+    SubscriberPort_test()
+    {
+    }
+
+    ~SubscriberPort_test()
+    {
+    }
+
+    void SetUp()
+    {
+    }
+
+    void TearDown()
+    {
+    }
+
+    iox::popo::SubscriberPortData m_subscriberPortDataSingleProducer{
+        iox::capro::ServiceDescription("x", "y", "z"),
+        "myApp",
+        iox::cxx::VariantQueueTypes::SoFi_SingleProducerSingleConsumer};
+    iox::popo::SubscriberPortUser m_sutUserSideSingleProducer{&m_subscriberPortDataSingleProducer};
+    iox::popo::SubscriberPortSingleProducer m_sutRouDiSideSingleProducer{&m_subscriberPortDataSingleProducer};
+};
+
+TEST_F(SubscriberPort_test, initialStateNotSubscribed)
+{
+    EXPECT_THAT(m_sutUserSideSingleProducer.getSubscriptionState(), Eq(iox::SubscribeState::NOT_SUBSCRIBED));
+}
+
+TEST_F(SubscriberPort_test, initialStateNoChunksAvailable)
+{
+    auto maybeChunk = m_sutUserSideSingleProducer.getChunk();
+
+    EXPECT_FALSE(maybeChunk.has_error());
+    EXPECT_FALSE(maybeChunk.get_value().has_value());
+    EXPECT_FALSE(m_sutUserSideSingleProducer.hasNewChunks());
+}
+
+TEST_F(SubscriberPort_test, initialStateNoChunksLost)
+{
+    EXPECT_FALSE(m_sutUserSideSingleProducer.hasLostChunks());
+}
+
+TEST_F(SubscriberPort_test, initialStateReturnsNoCaProMessage)
+{
+    auto maybeCaproMessage = m_sutRouDiSideSingleProducer.getCaProMessage();
+
+    EXPECT_FALSE(maybeCaproMessage.has_value());
+}
+
+TEST_F(SubscriberPort_test, subscribeCallResultsInSubCaProMessage)
```

Comment:
What happens when we call `subscribe` multiple times? If we call it 5 times do we then receive 5 capro messages? Maybe we should test the current behavior...

Reply:
No, this is no state change. We store the subscribe request. If you call multiple times subscribe this has no influence on the request